lib/os/sem.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <misc/sem.h>
+#include <syscall_handler.h>
+
+#ifdef CONFIG_USERSPACE
+int sys_sem_init(struct sys_sem *sem, unsigned int initial_count,
+		unsigned int limit)
+{
+	if (sem == NULL || limit == 0 ||
+	    initial_count > limit || limit > INT_MAX) {
+		return -EINVAL;
+	}
+
+	atomic_set(&sem->futex.val, initial_count);
+	sem->limit = limit;
+
+	return 0;
+}
+
+unsigned int sys_sem_count_get(struct sys_sem *sem)
+{
+	int value = atomic_get(&sem->futex.val);
+
+	return value > 0 ? value : 0;
+}
+
+int sys_sem_give(struct sys_sem *sem)
+{
+	int ret = 0;
+	int old_value;
+
+	do {
+		old_value = atomic_get(&sem->futex.val);
+		if (old_value >= sem->limit) {
+			break;
+		}
+	} while (atomic_cas(&sem->futex.val, old_value,
+			old_value < 0 ? 1 : old_value + 1) == 0);
+
+	if (old_value < 0) {
+		ret = k_futex_wake(&sem->futex, true);
+
+		if (ret > 0) {
+			return 0;
+		}
+	} else if (old_value >= sem->limit) {
+		return -EAGAIN;
+	}
+
+	return ret;
+}
+
+int sys_sem_take(struct sys_sem *sem, s32_t timeout)
+{
+	int ret = 0;
+	int old_value;
+
+	do {
+		do {
+			old_value = atomic_get(&sem->futex.val);
+			if (old_value < 0) {
+				break;
+			}
+		} while (atomic_cas(&sem->futex.val,
+				old_value, old_value - 1) == 0);
+
+		if (old_value > 0) {
+			return 0;
+		}
+
+		ret = k_futex_wait(&sem->futex, -1, timeout);
+	} while (ret == 0 || ret == -EAGAIN);
+
+	return ret;
+}
+#else
```

Comment:
FWIW: I think this is backwards.  **This** is the cleaner, more straightforward and more futureproof synchronization primitive.  What we really want is for the legacy k_sem implementation to be reworked in terms of sys_sem, not the reverse.  And yes, I know that k_futex APIs are only defined when USERSPACE is, but that's a mistake too -- they're just minimal wrappers around an atomic_t and a wait_q and there's no particular reason we couldn't make them work in kernel code too.

Not an issue for this patch, but long term I think we want just one set of shared memory synchronization primitives.

Reply:
> they're just minimal wrappers around an atomic_t and a wait_q and there's no particular reason we couldn't make them work in kernel code too.

One thing at a time, this is brand new immature code and we need to make sure it is stable and correct before we go in and throw out our tried-and-tested k_sem implementation

Also to be honest: the whole point of this stuff is to avoid system calls. We have no system calls if user mode is disabled.

Reply:
> there's no particular reason we couldn't make them work in kernel code too.

Actually there is: there is a dependency on the kernel object tracking for the futexes to instantiate the backing waitqueue object.

We could conditionally just add the waitq to the futex data struct if userspace is disabled, but then we have to worry about waitq initialization; currently futexes do not need to be initialized before they are used.

What I'm trying to say here is that this is not as simple as you think. I still want to do this, but not until at least 2.1